django/forms/formsets.py

```diff
@@ -105,6 +105,33 @@ def management_form(self):
             })
         return form
 
+    def _extract_form_indices(self):
+        """Find indices of all submitted forms"""
+        prefix = self.prefix + '-'
+        prefix_len = len(prefix)
+        for key in self.data.keys():
+            if key.startswith(prefix):
+                identifier = key[prefix_len:].split('-', 1)[0]
+                if identifier.isdigit():
+                    yield int(identifier)
+
+    def _extract_total_form_count(self):
+        """Extract the number of submitted forms"""
+        if self.management_form.cleaned_data[TOTAL_FORM_COUNT]:
+            # prefer the management form if data is present
+            return self.management_form.cleaned_data[TOTAL_FORM_COUNT]
+        try:
+            return max(self._extract_form_indices()) + 1
+        except ValueError:  # no forms submitted
+            return 0
+
+    def _extract_initial_form_count(self):
+        """Extract the number of initial forms"""
+        if self.management_form.cleaned_data[INITIAL_FORM_COUNT]:
+            # prefer the management form if data is present
+            return self.management_form.cleaned_data[INITIAL_FORM_COUNT]
+        return 0
```

Comment:
I don't think the formset and form situations are parallel. Initial form count is used in a variety of ways throughout formsets and model formsets, including when bound. A probably-incomplete list, gleaned from a quick grep:

* Deciding whether a form should be permitted/ignored when totally empty/unchanged (this is allowed for extra/additional/new forms, but not for initial forms).
* Many uses in inline-formsets, including deciding whether or not to look for a PK field, deciding whether to pass an `instance` to the individual form when instantiating it, deciding which forms should get initial data passed to them from the initial data passed to the inline formset.

I am extremely skeptical that it is feasible (or worthwhile) to correctly maintain backwards-compatible behavior in all these cases while removing the concept of initial forms.